File: android-app/core/core-network/src/main/java/tv/bayit/plus/core/network/NetworkConfiguration.kt
package tv.bayit.plus.core.network

import kotlin.time.Duration

/**
 * Configuration for the Bayit+ networking layer.
 *
 * All values are injected from the host application's configuration system
 * (BuildConfig, remote config, etc.) -- no defaults are embedded here.
 *
 * Mirrors the iOS [NetworkConfiguration] protocol from BayitNetworking.
 *
 * Implements [NetworkConfig] for backward compatibility.
 */
data class NetworkConfiguration(
    /** Base URL for all API requests (e.g. "https://api.bayit.tv/api/v1"). */
    override val baseUrl: String,

    /** Base URL for WebSocket connections (e.g. "wss://ws.bayit.tv"). */
    override val webSocketBaseUrl: String,

    /** Timeout for connect, read, and write operations. */
    val timeoutDuration: Duration,

    /** Maximum number of automatic retries for retryable failures. */
    override val maxRetries: Int,

    /** Base delay for the first retry. Subsequent retries use exponential backoff. */
    val retryBaseDelayDuration: Duration,

    /**
     * HTTP status codes eligible for automatic retry.
     * Matches the web api.js set: 408, 429, 500, 502, 503, 504.
     */
    val retryableStatusCodes: Set<Int>,

    /** Maximum number of concurrent WebSocket connections allowed. */
    override val webSocketMaxConnections: Int,

    /** Interval between WebSocket ping/keepalive messages. */
    val webSocketPingIntervalDuration: Duration,

    /** Maximum number of WebSocket reconnection attempts before giving up. */
    val webSocketMaxReconnectAttempts: Int,

    /** Base delay for WebSocket reconnection backoff. */
    val webSocketReconnectBaseDelay: Duration,

    /** Whether the app is in debug mode. Controls HTTP logging verbosity. */
    val isDebug: Boolean,
) : NetworkConfig {

    /** Timeout in milliseconds for OkHttp configuration. */
    val timeoutMillis: Long get() = timeoutDuration.inWholeMilliseconds

    /** Retry base delay in milliseconds for interceptor calculations. */
    val retryBaseDelayMillis: Long get() = retryBaseDelayDuration.inWholeMilliseconds

    /** WebSocket ping interval in seconds for OkHttp configuration. */
    val webSocketPingIntervalSeconds: Long get() = webSocketPingIntervalDuration.inWholeSeconds

    // NetworkConfig interface implementations (backward compatibility)
    override val timeout: Long get() = timeoutMillis
    override val retryBaseDelay: Long get() = retryBaseDelayMillis
    override val webSocketPingInterval: Long get() = webSocketPingIntervalSeconds
}

File: ios-app/Packages/BayitNetworking/Sources/BayitNetworking/NetworkConfiguration.swift
import Foundation

/// Protocol defining all configurable values for the networking layer.
///
/// No defaults are provided -- the host app MUST supply every value
/// from its configuration system (Info.plist, environment, remote config).
public protocol NetworkConfiguration: Sendable {
    /// Base URL for all API requests (e.g. `https://api.bayit.tv/api/v1`).
    var baseURL: URL { get }

    /// URLSession timeout interval for each request, in seconds.
    var timeout: TimeInterval { get }

    /// Maximum number of automatic retries for retryable failures.
    var maxRetries: Int { get }

    /// Base delay in seconds for the first retry. Subsequent retries
    /// use exponential back-off: `retryBaseDelay * 2^retryCount`.
    var retryBaseDelay: TimeInterval { get }

    /// HTTP status codes that are eligible for automatic retry.
    /// Matches the web api.js set: 408, 429, 500, 502, 503, 504.
    var retryableStatusCodes: Set<Int> { get }

    /// Default HTTP headers appended to every request.
    /// Content-Type is always set to `application/json` by the client,
    /// but additional static headers can be injected here.
    var defaultHeaders: [String: String] { get }

    // MARK: - WebSocket Configuration

    /// Maximum number of concurrent WebSocket connections allowed.
    var webSocketMaxConcurrentConnections: Int { get }

    /// Interval in seconds between WebSocket ping/keepalive messages.
    var webSocketPingInterval: TimeInterval { get }

    /// Maximum number of reconnection attempts before giving up.
    var webSocketMaxReconnectAttempts: Int { get }

    /// Base delay in seconds for WebSocket reconnection backoff.
    var webSocketReconnectBaseDelay: TimeInterval { get }

    /// Grace period in seconds before disconnecting WebSockets when app becomes inactive.
    var webSocketInactiveGracePeriod: TimeInterval { get }

    /// Base URL for WebSocket connections (e.g. `wss://ws.bayit.tv`).
    var webSocketBaseURL: URL { get }
}
